File: server/services/studyplan_service.go
package services

import (
	"context"
	"errors"
	"time"

	"buddy-server/database"
	"buddy-server/models"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

// StudyPlanService handles study plan operations
type StudyPlanService struct {
	db *database.DB
}

// NewStudyPlanService creates a new study plan service
func NewStudyPlanService(db *database.DB) *StudyPlanService {
	return &StudyPlanService{db: db}
}

// CreateStudyPlan creates a new study plan
func (s *StudyPlanService) CreateStudyPlan(userID, name, description string, startDate, endDate time.Time, dailyGoalHours float64, isChallenge, isPublic bool) (*models.StudyPlan, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	objectID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		return nil, errors.New("invalid user ID")
	}

	plan := &models.StudyPlan{
		UserID:         objectID,
		Name:           name,
		Description:    description,
		StartDate:      startDate,
		EndDate:        endDate,
		DailyGoalHours: dailyGoalHours,
		IsChallenge:    isChallenge,
		IsPublic:       isPublic,
		Progress:       0,
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
	}

	collection := s.db.Collection("study_plans")
	result, err := collection.InsertOne(ctx, plan)
	if err != nil {
		return nil, err
	}

	plan.ID = result.InsertedID.(primitive.ObjectID)
	return plan, nil
}

// GetStudyPlans gets all study plans for a user
func (s *StudyPlanService) GetStudyPlans(userID string) ([]models.StudyPlan, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	objectID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		return nil, errors.New("invalid user ID")
	}

	collection := s.db.Collection("study_plans")
	cursor, err := collection.Find(ctx, bson.M{"user_id": objectID})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var plans []models.StudyPlan
	if err = cursor.All(ctx, &plans); err != nil {
		return nil, err
	}

	return plans, nil
}

// GetPublicStudyPlans gets all public study plans (challenges)
func (s *StudyPlanService) GetPublicStudyPlans() ([]models.StudyPlan, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	collection := s.db.Collection("study_plans")
	cursor, err := collection.Find(ctx, bson.M{"is_public": true, "is_challenge": true})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var plans []models.StudyPlan
	if err = cursor.All(ctx, &plans); err != nil {
		return nil, err
	}

	return plans, nil
}

// AddCourse adds a course to a study plan
func (s *StudyPlanService) AddCourse(planID, subject string, hoursAllocated float64) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return errors.New("invalid plan ID")
	}

	course := &models.Course{
		StudyPlanID:    planObjectID,
		Subject:        subject,
		HoursAllocated: hoursAllocated,
		HoursCompleted: 0,
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
	}

	collection := s.db.Collection("courses")
	_, err = collection.InsertOne(ctx, course)
	return err
}

// GetCourses gets courses for a study plan
func (s *StudyPlanService) GetCourses(planID string) ([]models.Course, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return nil, errors.New("invalid plan ID")
	}

	collection := s.db.Collection("courses")
	cursor, err := collection.Find(ctx, bson.M{"study_plan_id": planObjectID})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var courses []models.Course
	if err = cursor.All(ctx, &courses); err != nil {
		return nil, err
	}

	return courses, nil
}

// CreateScheduleBlock creates a schedule block
func (s *StudyPlanService) CreateScheduleBlock(planID string, dayOfWeek int, startTime, endTime, subject, topic, blockType string) (*models.ScheduleBlock, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return nil, errors.New("invalid plan ID")
	}

	block := &models.ScheduleBlock{
		StudyPlanID: planObjectID,
		DayOfWeek:   dayOfWeek,
		StartTime:   startTime,
		EndTime:     endTime,
		Subject:     subject,
		Topic:       topic,
		BlockType:   blockType,
		CreatedAt:   time.Now(),
	}

	collection := s.db.Collection("schedule_blocks")
	result, err := collection.InsertOne(ctx, block)
	if err != nil {
		return nil, err
	}

	block.ID = result.InsertedID.(primitive.ObjectID)
	return block, nil
}

// GetScheduleBlocks gets schedule blocks for a study plan
func (s *StudyPlanService) GetScheduleBlocks(planID string) ([]models.ScheduleBlock, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return nil, errors.New("invalid plan ID")
	}

	collection := s.db.Collection("schedule_blocks")
	cursor, err := collection.Find(
		ctx,
		bson.M{"study_plan_id": planObjectID},
		// Sort by day of week and start time
	)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var blocks []models.ScheduleBlock
	if err = cursor.All(ctx, &blocks); err != nil {
		return nil, err
	}

	return blocks, nil
}

// UpdateScheduleBlock updates a schedule block
func (s *StudyPlanService) UpdateScheduleBlock(planID, blockID string, data map[string]interface{}) (*models.ScheduleBlock, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return nil, errors.New("invalid plan ID")
	}

	blockObjectID, err := primitive.ObjectIDFromHex(blockID)
	if err != nil {
		return nil, errors.New("invalid block ID")
	}

	// Verify block belongs to plan
	collection := s.db.Collection("schedule_blocks")
	var existing models.ScheduleBlock
	err = collection.FindOne(ctx, bson.M{
		"_id":          blockObjectID,
		"study_plan_id": planObjectID,
	}).Decode(&existing)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			return nil, errors.New("schedule block not found")
		}
		return nil, err
	}

	// Build update document
	update := bson.M{}
	if dayOfWeek, ok := data["day_of_week"]; ok {
		if d, ok := dayOfWeek.(float64); ok {
			update["day_of_week"] = int(d)
		} else if d, ok := dayOfWeek.(int); ok {
			update["day_of_week"] = d
		}
	}
	if startTime, ok := data["start_time"]; ok {
		if s, ok := startTime.(string); ok {
			update["start_time"] = s
		}
	}
	if endTime, ok := data["end_time"]; ok {
		if e, ok := endTime.(string); ok {
			update["end_time"] = e
		}
	}
	if subject, ok := data["subject"]; ok {
		if s, ok := subject.(string); ok {
			update["subject"] = s
		}
	}
	if topic, ok := data["topic"]; ok {
		if t, ok := topic.(string); ok {
			update["topic"] = t
		}
	}
	if blockType, ok := data["block_type"]; ok {
		if b, ok := blockType.(string); ok {
			update["block_type"] = b
		}
	}

	if len(update) == 0 {
		return &existing, nil // No updates
	}

	// Update block
	_, err = collection.UpdateOne(
		ctx,
		bson.M{"_id": blockObjectID},
		bson.M{"$set": update},
	)
	if err != nil {
		return nil, err
	}

	// Fetch updated block
	var updated models.ScheduleBlock
	err = collection.FindOne(ctx, bson.M{"_id": blockObjectID}).Decode(&updated)
	if err != nil {
		return nil, err
	}

	return &updated, nil
}

// GetUserSchedule gets all schedule blocks for a user across all plans
func (s *StudyPlanService) GetUserSchedule(userID string) ([]models.ScheduleBlock, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userObjectID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		return nil, errors.New("invalid user ID")
	}

	// First get user's study plans
	plansCollection := s.db.Collection("study_plans")
	cursor, err := plansCollection.Find(ctx, bson.M{"user_id": userObjectID})
	if err != nil {
		return nil, err
	}

	var plans []models.StudyPlan
	if err = cursor.All(ctx, &plans); err != nil {
		return nil, err
	}
	cursor.Close(ctx)

	// Get plan IDs
	var planIDs []primitive.ObjectID
	for _, plan := range plans {
		planIDs = append(planIDs, plan.ID)
	}

	if len(planIDs) == 0 {
		return []models.ScheduleBlock{}, nil
	}

	// Get all schedule blocks for these plans
	blocksCollection := s.db.Collection("schedule_blocks")
	cursor, err = blocksCollection.Find(ctx, bson.M{
		"study_plan_id": bson.M{"$in": planIDs},
	})
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var blocks []models.ScheduleBlock
	if err = cursor.All(ctx, &blocks); err != nil {
		return nil, err
	}

	return blocks, nil
}

// UpdateStudyPlanProgress updates the progress of a study plan
func (s *StudyPlanService) UpdateStudyPlanProgress(planID string, progress float64) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	planObjectID, err := primitive.ObjectIDFromHex(planID)
	if err != nil {
		return errors.New("invalid plan ID")
	}

	collection := s.db.Collection("study_plans")
	_, err = collection.UpdateOne(
		ctx,
		bson.M{"_id": planObjectID},
		bson.M{"$set": bson.M{
			"progress":   progress,
			"updated_at": time.Now(),
		}},
	)

	return err
}

File: desktop/frontend/src/services/backend/studyplan.service.ts
import { supabase } from './auth.service';

export interface StudyPlan {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  start_date: string;
  end_date: string;
  daily_goal_hours: number;
  is_challenge: boolean;
  is_public: boolean;
  progress: number;
  created_at: string;
  updated_at: string;
}

export interface StudyPlanCourse {
  id: string;
  study_plan_id: string;
  subject: string;
  hours_allocated: number;
  hours_completed: number;
}

export interface ScheduleBlock {
  id: string;
  study_plan_id: string;
  day_of_week: number;
  start_time: string;
  end_time: string;
  subject: string;
  topic: string;
  block_type: 'study' | 'practice' | 'review' | 'test';
}

export interface CreateStudyPlanData {
  user_id: string;
  name: string;
  description?: string;
  start_date: string;
  end_date: string;
  daily_goal_hours: number;
  is_challenge?: boolean;
  is_public?: boolean;
}

export const studyPlanService = {
  async createStudyPlan(data: CreateStudyPlanData) {
    const { data: plan, error } = await supabase
      .from('study_plans')
      .insert({
        ...data,
        progress: 0,
        is_challenge: data.is_challenge || false,
        is_public: data.is_public || false,
      })
      .select()
      .single();

    if (error) throw error;
    return plan as StudyPlan;
  },

  async getStudyPlans(userId: string) {
    const { data, error } = await supabase
      .from('study_plans')
      .select(`
        *,
        study_plan_courses (*)
      `)
      .eq('user_id', userId)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return data;
  },

  async getStudyPlan(planId: string) {
    const { data, error } = await supabase
      .from('study_plans')
      .select(`
        *,
        study_plan_courses (*),
        schedule_blocks (*)
      `)
      .eq('id', planId)
      .maybeSingle();

    if (error) throw error;
    return data;
  },

  async updateStudyPlan(planId: string, updates: Partial<StudyPlan>) {
    const { data, error } = await supabase
      .from('study_plans')
      .update({
        ...updates,
        updated_at: new Date().toISOString(),
      })
      .eq('id', planId)
      .select()
      .single();

    if (error) throw error;
    return data as StudyPlan;
  },

  async deleteStudyPlan(planId: string) {
    const { error } = await supabase
      .from('study_plans')
      .delete()
      .eq('id', planId);

    if (error) throw error;
  },

  async addCourse(planId: string, subject: string, hoursAllocated: number) {
    const { data, error } = await supabase
      .from('study_plan_courses')
      .insert({
        study_plan_id: planId,
        subject,
        hours_allocated: hoursAllocated,
        hours_completed: 0,
      })
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  async updateCourseProgress(courseId: string, hoursCompleted: number) {
    const { data, error } = await supabase
      .from('study_plan_courses')
      .update({ hours_completed: hoursCompleted })
      .eq('id', courseId)
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  async addScheduleBlock(block: Omit<ScheduleBlock, 'id'>) {
    const { data, error } = await supabase
      .from('schedule_blocks')
      .insert(block)
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  async deleteScheduleBlock(blockId: string) {
    const { error } = await supabase
      .from('schedule_blocks')
      .delete()
      .eq('id', blockId);

    if (error) throw error;
  },

  async getPublicStudyPlans() {
    const { data, error } = await supabase
      .from('study_plans')
      .select(`
        *,
        profiles (
          full_name,
          avatar_url
        )
      `)
      .eq('is_public', true)
      .eq('is_challenge', true)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return data;
  },

  async joinChallenge(planId: string, userId: string) {
    const { data, error } = await supabase
      .from('study_plan_participants')
      .insert({
        study_plan_id: planId,
        user_id: userId,
      })
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  async getChallengeParticipants(planId: string) {
    const { data, error } = await supabase
      .from('study_plan_participants')
      .select(`
        *,
        profiles (
          full_name,
          avatar_url
        )
      `)
      .eq('study_plan_id', planId);

    if (error) throw error;
    return data;
  },
};
